src/dmd/optimize.d

```diff
@@ -432,7 +432,7 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
                 return;
             }
             // Keep lvalue-ness
-            if (expOptimize(e.e1, result, true))
+            if (e.e1.op != TOK.cast_ && expOptimize(e.e1, result, true))
```

Comment:
I guess the problem lies deeper. As the comment says, the `true` param stands for `keepLvalue`, but it clearly isn't kept. Btw, that `keepLvalue` should almost certainly be set in the `optimize()`-call touched in #9505 (that was my attempt, but wasn't enough due to `keepLvalue` apparently not properly being respected here in this file).

Reply:
Yes, I noticed that the comment is not in sync with what actually happens, but I thought that maybe it referred to something else. It is a bit puzzling that the code that does optimization rewrittes without too much thought lvalues into rvalues.

Reply:
This particular test case (`&(cast() x)`) seems to be fixed by this patch for `OptimizeVisitor` visiting a `CastExp`:
```diff
@@ -625,9 +625,10 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
             assert(e.type);
             TOK op1 = e.e1.op;
             Expression e1old = e.e1;
-            if (expOptimize(e.e1, result))
+            if (expOptimize(e.e1, result, keepLvalue))
                 return;
-            e.e1 = fromConstInitializer(result, e.e1);
+            if (!keepLvalue)
+                e.e1 = fromConstInitializer(result, e.e1);
             if (e.e1 == e1old && e.e1.op == TOK.arrayLiteral && e.type.toBasetype().ty == Tpointer && e.e1.type.toBasetype().ty != Tsarray)
             {
                 // Casting this will result in the same expression, and
```

---

In combination with this follow-up to #9505:
```diff
@@ -8749,8 +8749,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             if (exp.op == TOK.assign)
                 e1x = e1x.modifiableLvalue(sc, e1old);

-            if (e1x.op != TOK.variable)
-                e1x = e1x.optimize(WANTvalue);
+            import core.stdc.stdio;
+            printf(".: e1x before optimize(): %s\n", e1x.toChars());
+            e1x = e1x.optimize(WANTvalue);//, /*keepLvalue*/ true);
+            printf(".: e1x after  optimize(): %s\n", e1x.toChars());

             if (e1x.op == TOK.error)
             {
```

it now does what I expect for `const x = 0; (cast() x) = 5;` (`before: cast(int)x, after: x`, previously wrongly const-folded to `after: 0`) - but LDC's final codegen-AST remains `0 = 5;`. :/

Reply:
Ah, the `AssignExp` undergoes another `optimize()` pass later on, which exposes another issue in the `OptimizeVisitor`. I'll open another PR.

Reply:
=> #10124